### PR TITLE
[Performance] Add NCCL support 

### PR DESCRIPTION
**Pull request by isratnisa**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by yaox12 — APPROVED** [review]
It seems we have to check `torch.distributed.is_initialized()` before calling `th.distributed.get_backend()`. Otherwise LGTM,

File: python/dgl/distributed/optim/pytorch/sparse_optim.py [hunk 1]
```diff
@@ -323,7 +327,11 @@ def step(self):
                                 idx_j = idx_i[mask]
                                 grad_j = grad_i[mask]
                                 idx_split_size.append(
-                                    th.tensor([idx_j.shape[0]], dtype=th.int64)
+                                    th.tensor(
+                                        [idx_j.shape[0]],
+                                        dtype=th.int64,
+                                        device=device,
```
Comment (Rhett-Ying) [inline python/dgl/distributed/optim/pytorch/sparse_optim.py:333]: what if multiple GPUs are being used during train, do we need to specify device id like `cuda:0/1/2/3` here?

File: python/dgl/distributed/optim/pytorch/sparse_optim.py [hunk 2]
```diff
@@ -256,9 +256,9 @@ def step(self):
         of the embeddings involved in a mini-batch to DGL's servers and update the embeddings.
         """
         with th.no_grad():
+            device = "cuda" if th.distributed.get_backend() == "nccl" else "cpu"
```
Comment (Rhett-Ying) [inline python/dgl/distributed/optim/pytorch/sparse_optim.py:259]: we'd better wrap with `torch.device()`?

File: python/dgl/distributed/optim/pytorch/sparse_optim.py [hunk 3]
```diff
@@ -336,40 +344,65 @@ def step(self):
...
-                    alltoall_cpu(
-                        self._rank,
-                        self._world_size,
-                        gather_list,
-                        idx_split_size,
+                        th.empty(
+                            [self._world_size], dtype=th.int64, device=device
+                        ).chunk(self._world_size)
                     )
+                    if th.distributed.get_backend() == "nccl":
```
Comment (Rhett-Ying) [inline python/dgl/distributed/optim/pytorch/sparse_optim.py:351]: how about wrapping these if-else into a function like `dgl.all_to_all()` which is more brief in code or you need to add such logic in multiple places.

**Review by Rhett-Ying — CHANGES_REQUESTED** [review]
dedicated unit test is required

File: python/dgl/distributed/optim/pytorch/sparse_optim.py (hunk 1 above)
Comment (isratnisa) [reply python/dgl/distributed/optim/pytorch/sparse_optim.py:334]: I think you are right. Changed device to   `device = th.device(f"cuda:{self._rank}")`

File: python/dgl/distributed/optim/pytorch/sparse_optim.py (hunk 2 above)
Comment (isratnisa) [reply python/dgl/distributed/optim/pytorch/sparse_optim.py]: Fixed

File: python/dgl/distributed/optim/pytorch/sparse_optim.py (hunk 3 above)
Comment (isratnisa) [reply python/dgl/distributed/optim/pytorch/sparse_optim.py]: Added.

**Review by Rhett-Ying — COMMENTED** [review]
could you add test cases into https://github.com/dmlc/dgl/blob/master/tests/python/pytorch/distributed/optim/test_dist_optim.py? and please resolve the lint error first.

**Review by Rhett-Ying — APPROVED** [review]
please fix the lint issue before merge

**Pull request merged**